src/python/gudhi/tensorflow/cubical_layer.py

```diff
@@ -0,0 +1,72 @@
+import numpy               as np
+import tensorflow          as tf
+from ..cubical_complex  import CubicalComplex
+
+######################
+# Cubical filtration #
+######################
+
+# The parameters of the model are the pixel values.
+
+def _Cubical(X, dimensions):
+    # Parameters: X (image),
+    #             dimensions (homology dimensions)
+
+    # Compute the persistence pairs with Gudhi
+    Xs = X.shape
+    cc = CubicalComplex(dimensions=Xs, top_dimensional_cells=X.flatten())
+    cc.persistence()
+
+    L_cofs = []
+    for dim in dimensions:
+
+        try:
+            cof = cc.cofaces_of_persistence_pairs()[0][dim]
+        except IndexError:
+            cof = np.array([])
+
+        # Retrieve and ouput image indices/pixels corresponding to positive and negative simplices
+        D = len(Xs) if len(cof) > 0 else 1
+        ocof = np.array([0 for _ in range(D*2*cof.shape[0])])
+        count = 0
+        for idx in range(0,2*cof.shape[0],2):
+            ocof[D*idx:D*(idx+1)]     = np.unravel_index(cof[count,0], Xs)
```

Comment:
If I understand correctly, cofaces_of_persistence_pairs returns a flat index, which you unravel, so you can index in the original tensor. Would it be possible instead to flatten the tensorflow index and use directly the flat index in it?

Reply:
Good catch! Indeed there was an issue when I did a test with a non-square array, so I made sure the order is right with the option `order='F'` in `np.unravel_index`

Reply:
Could you add that non-square test, to ensure we don't break it later?
What do you think of the option of not unraveling?

Reply:
Yep, I added the non-square test. I tried flattening, but since tensorflow does not use Fortran ordering, I had to transpose the matrix first. All in all, I think the code is simpler that way, so I think we should keep this version (with flat indices). Thanks Marc!

Reply:
Note that there is an alternative to transposing the matrix: mirroring the dimensions. If the dimensions are `(a,b,c)` and you give CubicalComplex `(c,b,a)` and `matrix.flatten()`, it *should* also work :crossed_fingers: .

Reply:
Even better! Thanks Marc, now the only operation we do is a flatten at the beginning for initializing the CubicalComplex.